packages/core/contracts/financial-templates/funding-rate-store/implementation/OptimisticOracle.sol

```diff
@@ -0,0 +1,357 @@
+import "../../../oracle/implementation/Constants.sol";
+
+import "../../../common/implementation/Testable.sol";
+import "../../../common/implementation/Lockable.sol";
+import "../../../common/implementation/FixedPoint.sol";
+import "../../../common/implementation/AddressWhitelist.sol";
+
+interface FundingRateRequester {
+    function priceProposed(bytes32 identifier, uint256 timestamp) external;
+
+    function priceDisputed(
+        bytes32 identifier,
+        uint256 timestamp,
+        uint256 refund
+    ) external;
+
+    function priceSettled(
+        bytes32 identifier,
+        uint256 timestamp,
+        int256 price
+    ) external;
+}
+
+contract OptimisticOracle is Testable, Lockable {
+    using SafeMath for uint256;
+    using SafeERC20 for IERC20;
+
+    event RequestPrice(address indexed requester, bytes32 identifier, uint256 timestamp);
+    event ProposePrice(
+        address indexed requester,
+        address indexed proposer,
+        bytes32 identifier,
+        uint256 timestamp,
+        int256 proposedPrice
+    );
+    event DisputePrice(
+        address indexed requester,
+        address indexed proposer,
+        address indexed disputer,
+        bytes32 identifier,
+        uint256 timestamp
+    );
+    event Settle(
+        address indexed requester,
+        address indexed proposer,
+        address indexed disputer,
+        bytes32 identifier,
+        uint256 timestamp,
+        int256 price
+    );
+
+    enum State {
+        Invalid, // Never requested.
+        Requested, // Requested, no other actions taken.
+        Proposed, // Proposed, but not expired or disputed yet.
+        Expired, // Proposed, not disputed, past liveness.
+        Disputed, // Disputed, but no DVM price returned yet.
+        Resolved, // Disputed and DVM price is available.
+        Settled // Final price has been set in the contract (can get here from Expired or Resolved).
+    }
+
+    struct Request {
+        address proposer;
+        address disputer;
+        IERC20 currency;
+        bool settled;
+        bool refundOnDispute;
+        int256 proposedPrice;
+        int256 resolvedPrice;
+        uint256 expirationTime;
+        uint256 reward;
+        uint256 finalFee;
+        uint256 bond;
+        uint256 customLiveness;
+    }
+
+    mapping(bytes32 => Request) public requests;
+
+    FinderInterface public finder;
+    uint256 public defaultLiveness;
+
+    constructor(
+        address _finderAddress,
+        address _timerAddress,
+        uint256 _liveness
+    ) public Testable(_timerAddress) {
+        finder = FinderInterface(_finderAddress);
+        _validateLiveness(_liveness);
+        defaultLiveness = _liveness;
+    }
+
+    function requestPrice(
+        bytes32 identifier,
+        uint256 timestamp,
+        IERC20 currency,
+        uint256 reward
+    ) external nonReentrant() returns (uint256 totalBond) {
+        bytes32 id = _getId(msg.sender, identifier, timestamp);
```

Comment:
nit: put this below the requires as none of them need it.

Reply:
in fact, this is more of a stylistic thing than anything else. there are a few cases in this contract where you could put the require's above any local variable decleration.

Reply:
Good point. This was leftover from some previous churn, definitely agree that it should go below the requires.

Reply:
Done.